### PR TITLE
Adding adobe form support, part 2

### PR DESCRIPTION
**Pull request by afrosimon**
Couldn't modify the other pull request so I had to create a new one. As before let me know if something is off.

### CODE REVIEW

File: share/pdf2htmlEX.js.in [hunk 1]
```diff
@@ -260,7 +260,24 @@ Viewer.prototype = {
     this.pre_hide_pages();
   },
 
+  initialize_radio_button : function() {
+    var elements = document.getElementsByClassName('ir');
+    
+    for(var i = 0; i < elements.length; i++) {
```
Comment (coolwanglu) [inline share/pdf2htmlEX.js.in]: how about using `document.getElementsByClassName(...).forEach` here?

File: share/pdf2htmlEX.js.in [hunk 2]
```diff
@@ -260,7 +260,24 @@ Viewer.prototype = {
...
+      var r = elements[i];
+
+      r.addEventListener('click', function() {
+        if(this.className.search("checked") == -1)
```
Comment (coolwanglu) [inline share/pdf2htmlEX.js.in]: this could be made simpler with `classList`

File: share/pdf2htmlEX.js.in [hunk 3]
```diff
@@ -260,7 +260,24 @@ Viewer.prototype = {
...
+          this.className += " checked";
+        else
+          this.className = "ir";
+      });
+    }
+  },
+
   init_after_loading_content : function() {
+    this.initialize_radio_button();
```
Comment (coolwanglu) [inline share/pdf2htmlEX.js.in]: Please move this near the end of this function.

File: src/HTMLRenderer/form.cc [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+void HTMLRenderer::process_form(ofstream & out)
+{
+    FormPageWidgets * widgets = cur_catalog->getPage(pageNum)->getFormWidgets();
+    int num = widgets->getNumWidgets();
+
+    for(int i = 0; i < num; i++)
+    {
+        FormWidget * w = widgets->getWidget(i);
+        double x1, y1, x2, y2, width, font_size;
```
Comment (coolwanglu) [inline src/HTMLRenderer/form.cc]: Please declare variable near the first usage.

File: src/HTMLRenderer/form.cc [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+        width = x2 - x1;
+        height = y2 - y1;
+        
+        if(w->getType() == formText)
+        {
+            font_size = height / 2;
+
+            out 
+                << "<input id=\"text-" << pageNum << "-"
```
Comment (coolwanglu) [inline src/HTMLRenderer/form.cc]: Please rearrange this part, it looks messy here.
Also please move `class` to right after `id`.

Example: https://github.com/coolwanglu/pdf2htmlEX/blob/master/src/HTMLTextLine.cc#L151

File: src/HTMLRenderer/form.cc [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+                    << CSS::INPUT_TEXT_CN << "\" />" << endl;
+        }
+
+        if(w->getType() == formButton)
+        {
+            width += 3;
+            height += 3;
+
+            out 
+                << "<div id=\"cb-" << pageNum << "-" << i << "\""
```
Comment (coolwanglu) [inline src/HTMLRenderer/form.cc]: same as above.

File: src/HTMLRenderer/form.cc [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+                    << i << "\" type=\"text\" value=\"\""
+                    << " style=\"position: absolute; left: " << x1 
+                    << "px; bottom: " << y1 << "px;" 
+                    << " width: " << width << "px; height: " << std::to_string(height) 
+                    << "px; line-height: " << std::to_string(height) << "px; font-size: " 
+                    << font_size << "px;\" class=\"" 
+                    << CSS::INPUT_TEXT_CN << "\" />" << endl;
+        }
+
+        if(w->getType() == formButton)
```
Comment (coolwanglu) [inline src/HTMLRenderer/form.cc]: use `else if` here.

File: src/HTMLRenderer/form.cc [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+            out 
+                << "<div id=\"cb-" << pageNum << "-" << i << "\"" 
+                    << " style=\"position: absolute; left: " << x1 
+                    << "px; bottom: " << y1 << "px;" 
+                    << " width: " << width << "px; height: " 
+                    << std::to_string(height) << "px; background-size: cover;\" class=\"" 
+                    << CSS::INPUT_RADIO_CN << "\"></div>" << endl;
+
+        }
```
Comment (coolwanglu) [inline src/HTMLRenderer/form.cc]: If there are other types of forms in PDF, please add one more `else` clause here, and print some warning message about unsupported form type.

**Comment by coolwanglu** [comment]
Thanks for the PR! It looks in good shape to me, just need a few more modifications.

It seems that Travis CI failed with this patch, can you find the reason? Are you able to run the test cases locally?

File: share/pdf2htmlEX.js.in (hunk 1 above)
Comment (afrosimon) [inline share/pdf2htmlEX.js.in]: Tried it, we can't on an HTMLCollection (http://stackoverflow.com/questions/24266313/why-does-my-call-to-foreach-on-an-array-result-in-uncaught-typeerror-undefined)

File: share/pdf2htmlEX.js.in (hunk 2 above)
Comment (afrosimon) [inline share/pdf2htmlEX.js.in]: Indeed... Did not know about classList. Thanks for the tip

**Comment by coolwanglu** [comment]
Oh, and don't forget to add yourself to `AUTHORS` :)

File: src/HTMLRenderer/form.cc (hunk 4 above)
Comment (afrosimon) [inline src/HTMLRenderer/form.cc]: I fixed the rest, but this one I do not understand. Should I have declared widgets  in the header file?

**Comment by afrosimon** [comment]
I'm unsure why the Travis build is failing, I'm not very familiar with this tool. Here's what I can tell you : 

 - I've always had an error on the test cases : "AssertionError: Screen is not large enough" Up until the test_local_browser part everything is fine. If I comment the specific line very test is fine, no errors whatsoever (i.e. python test.py)

 - As for Travis, here's what I noticed : If I compare the output with, say, the build #652, I can see I'm not exporting SAUCE_USERNAME nor the password (on the build #652 this is done around line 20, on my build this would have been done at around line 84). I find this puzzling as I did not touch the .travis.yml file. Anything else I could try?

**Comment by coolwanglu** [comment]
@afrosimon the `SAUCE` stuff is protected, so probably only I can test it, and I will do it.
So don't worry about the test cases, I'll take it here.

Please add yourself to `AUTHORS`

**Comment by afrosimon** [comment]
I've implemented the last few changes you've requested, and I've added myself to the authors file. Let me know if there's anything else!

File: src/HTMLRenderer/form.cc (hunk 4 above)
Comment (coolwanglu) [inline src/HTMLRenderer/form.cc]: I mean define the variable when it is used the first time.
For example, write `double width = x2-x1;` later, also please do this for `height` and `font_size`.

By the way, why `height` is int?

**Comment by afrosimon** [comment]
I've just cleaned up the variable declaration in form.cc, I believe that's about everything.

**Comment by coolwanglu** [comment]
Looks good to me now. Thanks for your effort!
I'll try to fix the tests.

**Comment by coolwanglu** [comment]
@afrosimon I found that the test case you provided triggers #483, I wonder if you can provided a simpler test case, just to fix the auto tests.

**Comment by afrosimon** [comment]
By simpler, you mean, with less images / text? As it stands most of the
documents I have are about all as complex as the sample provided. I'll
search for something simpler if that is what is needed, but I have to ask,
is it not the point of a test suite to find out such bugs?

**Comment by coolwanglu** [comment]
@afrosimon Yes, I will also use this file to test that bug. What I meant was to release a version before I fix this bug, as I didn't manage to solve it. 
But seems that I have a potential solution now, so a 'simpler' case is not necessary for now. I will ask you again if I cannot fix #483. Thanks.

**Pull request merged**